---
title: "Stromal fine annotation"
author: "Carla Cohen"
date: "`r Sys.Date()`"
output: html_document
---
# Stromal annotation  
A script to perform fine annotation of the adipocyte/VEC/LEC/mural subsets in the Achilles data.

This script follows the scVI integration of the whole Achilles dataset, subsetting, then reintegration with scVI 

Preliminary analysis of the clustering and multiple resolutions was performed: 20250114_15-14_Subset_Adipocyte.VEC.LEC.Mural.dir
This produced lists of DE genes and heatmaps. 

Now we need to perform fine annotation to see whether the subsets have any biological meaning. 

This script is for scVI louvain at res 0.5
AFter the highly ambient genes have been removed. 


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```


```{r}
library(tidyverse)
library(Seurat)
library(cowplot)
library(viridis)
library(scales)
library(readxl)
library(scDotPlot)
library(SingleCellExperiment)
library(ggsci)
library(data.table)
library(clustree)
library(scuttle)

# set the resolution for the script
resolution <- 0.5
clustering <- "louvain"

# make a new output folder for each run, with the date & time in the directory name
date <- Sys.Date() %>% str_replace_all("-", "")
time <- format(Sys.time(), "%X") %>% str_replace_all(":", "-") %>%
    str_sub(1, 5)
directory <- paste0(date,"_", time, "_Fine_annotation_stromal_", resolution, ".dir")
dir.create(directory, showWarnings = FALSE)
dir.create(paste0(directory, "/Marker_dotplots/"))
dir.create(paste0(directory, "/Annotation_figures/"))
dir.create(paste0(directory, "/Subset_figures/"))


# microanatomy colours
Okabe_Ito <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
ma.cols <-  c(Enth = Okabe_Ito[3], MB = Okabe_Ito[5], MTJ = Okabe_Ito[6], muscle = Okabe_Ito[7])
```

Read in the Achilles subset  object. 

```{r}
so.stromal <- readRDS("20250114_15-14_Subset_Adipocyte.VEC.LEC.Mural.dir/RDS_objects.dir/Achilles_Adipocyte.VEC.LEC.Mural_subset.rds")
```




# Resolution 0.5

Annotation at resolution 0.5

Select a resolution and set seurat clusters to that resolution 


```{r} 
so.stromal[["seurat_clusters"]] <- so.stromal[[paste0(clustering, "_X_scVI_snn_res.", resolution)]]
Idents(so.stromal) <- so.stromal$seurat_clusters
levels(Idents(so.stromal))
```



## Plot the UMAP at the chosen resolution  

```{r, fig.width=8, fig.height=6}
DimPlot(object = so.stromal, label = TRUE, reduction = "umap", shuffle = TRUE)+
    ggtitle(paste0("UMAP at resolution ", resolution))
ggsave(paste0(directory, "/Annotation_figures/Unlabelled_UMAP_", resolution, ".png"), 
       width = 6, height = 5, bg = "white")
```




UMAP split by microanatomy

```{r, fig.height=8, fig.width=8}
p <- DimPlot(object = so.stromal, label = TRUE, reduction = "umap", shuffle = TRUE, 
             split.by = "microanatomical_site", ncol = 2)+
    ggtitle(paste0("UMAP split my microanatomy at resolution ", resolution))
ggsave(paste0(directory, "/Annotation_figures/UMAP_by_microanatomy_", resolution, ".png"), 
       width = 8, height = 8, bg = "white")
p
```

> Take home from microanatomy

Hard to tell if real differences or just sampling. 
Clusters 8 and 9 appear to be only in MTJ/muscle

Bar chart of cell composition

```{r}

df <- so.stromal[[]] %>% select(seurat_clusters, microanatomical_site)

#library(data.table)
df <- data.table(df)
df <- df[, .(COUNT = .N), by = names(df)]

ggplot(df, aes(fill=microanatomical_site, y=COUNT, x=seurat_clusters)) + 
    geom_bar(position="fill", stat="identity")+
    theme_classic()+
    scale_fill_manual(values = ma.cols)

ggsave(paste0(directory, "/Annotation_figures/Cluster_fraction_microanatomy.png"), 
       width = 8, height = 6, bg = "white")
```


### UMAP split by patient

```{r, fig.height=8, fig.width=10}
DimPlot(object = so.stromal, reduction = "umap", shuffle = TRUE, 
             split.by = "patient", ncol = 3, label = TRUE)+
    theme(panel.background = element_rect(colour = "black", linewidth = 1))+
    ggtitle(paste0("UMAP split by patient at resolution ", resolution))
ggsave(paste0(directory, "/Annotation_figures/UMAP_patient.png"), 
       width = 10, height = 7, bg = "white")
```

### UMAP split by patient - each cluster separately

```{r, fig.height=6, fig.width=8}
# set the colours
colours <- hue_pal()(length(levels(Idents(so.stromal))))
names(colours) <- levels(Idents(so.stromal))

for (cluster_id in levels(Idents(so.stromal))){
    #print(cluster_id)
    so.subset <- subset(so.stromal, idents = cluster_id)
    DimPlot(so.subset, label = TRUE, reduction = "umap", shuffle = TRUE, 
             split.by = "patient", ncol = 3, cols = colours)+
    ggtitle(paste0("UMAP cluster ", cluster_id, " per patient res ", resolution))
    ggsave(paste0(directory, "/Annotation_figures/UMAP_cluster_", cluster_id, "_by_patient_", resolution, ".png"), 
       width = 8, height = 6, bg = "white")
}


```


### What % of nuclei are from each patient?

```{r}
# make a df of patient percentages per cluster
df <- so.stromal[[]] %>% select(seurat_clusters, patient)
df <- data.table(df)
df <- df[, .(COUNT = .N), by = names(df)]
df2 <- df %>% pivot_wider(names_from = patient, values_from = COUNT, values_fill = 0) %>% 
    column_to_rownames(var = "seurat_clusters")%>%
    mutate_at(vars(1:6), as.numeric) 
df2$my_row_sum <- rowSums(df2)
df2 <- df2 %>% mutate(across(where(is.numeric), ~ .x/my_row_sum*100))
df2$my_row_sum <- NULL
df2$seurat_clusters <- rownames(df2)

df2

# plot as bar chart and heatmap
df3 <- df2 %>% pivot_longer(1:6, names_to = "patient", values_to = "proportion")

ggplot(df3, aes(fill=patient, y=proportion, x=seurat_clusters)) + 
    geom_bar(position="stack", stat = "identity")+
    theme_classic()
ggsave(paste0(directory, "/Annotation_figures/Patient_proportion_barchart_", resolution, ".png"), 
       width = 8, height = 6, bg = "white")

ggplot(df3, aes(y=seurat_clusters, x=patient, fill = proportion))+
    geom_tile()+
    scale_fill_viridis()
ggsave(paste0(directory, "/Annotation_figures/Patient_proportion_heatmap_", resolution, ".png"), 
       width = 8, height = 6, bg = "white")


```

Cluster 8, 9 and 12 are skewed towards individual patients, proceed with caution. But none > 80%


### How many cells are there per cluster?

```{r}
cell_numbers <- table(Idents(so.stromal)) %>% as.data.frame()
colnames(cell_numbers) <- c("cluster", "frequency")
cell_numbers
```
```{r}
ggplot(cell_numbers, aes(x = cluster, y = frequency))+
    geom_bar(stat = "identity")+
    theme_cowplot()+
    ggtitle("Number of cells per cluster")
ggsave(paste0(directory, "/Annotation_figures/Number_cells_per_cluster_", resolution, ".png"), 
       width = 6, height = 4, bg = "white")
```


### Visualise QC metrics 

```{r, message = FALSE, fig.width=10, fig.height=18}
p1 <- VlnPlot(so.stromal, features = "sum", assay = "RNA", pt.size = 0)+
    scale_y_log10() +
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("nCounts")

p2 <- VlnPlot(so.stromal, features = "detected", assay = "RNA", pt.size = 0)+
    scale_y_log10() +
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("nFeatures")

p3 <- VlnPlot(so.stromal, features = "subsets_mito_percent", assay = "RNA", pt.size = 0)+
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("Percent mitochondrial reads")

p4 <- VlnPlot(so.stromal, features = "decontX_contamination", assay = "RNA")+
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("decontX score")

p5 <- VlnPlot(so.stromal, features = "percent.ribo", assay = "RNA")+
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("Percent ribosomal")

p6 <- VlnPlot(so.stromal, features = "soupX_fraction", assay = "RNA")+
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("SoupX fraction")

p <- plot_grid(p1, p2, p3, p4, p5, p6, ncol = 1)

ggsave(paste0(directory, "/Annotation_figures/QC_metrics_across_clusters.png"), 
       p, width = 10, height = 21, bg = "white")
p

```

Cluster 10 is high for ribosomal reads and ambient - likely Soup
Cluster 7 is also high for ambient. 
Cluster 5 is bimodal, could filter out the highly contaminating cells or subcluster further?

Plot QC metrics on UMAP

```{r, fig.height=12, fig.width=10, message = FALSE}
p1<- FeaturePlot(so.stromal, features = "sum", reduction = "umap")+ggtitle ("nCounts")+
    scale_color_continuous(type = "viridis", limits = c(0,10000))
p2<- FeaturePlot(so.stromal, features = "detected", reduction = "umap") + ggtitle("nFeatures")+
    scale_color_continuous(type = "viridis", limits = c(0,5000))
p3<- FeaturePlot(so.stromal, features = "subsets_mito_percent", reduction = "umap") + ggtitle ("Percent mito reads")+
    scale_color_continuous(type = "viridis")
p4<- FeaturePlot(so.stromal, features = "decontX_contamination", reduction = "umap")+
    scale_colour_gradientn(colours = viridis(256, option = "D"), name = "decontX score")+
        ggtitle("decontX_contamination")
p5 <- FeaturePlot(so.stromal, features = "percent.ribo", reduction = "umap")+
    scale_colour_gradientn(colours = viridis(256, option = "D"), name = "percent.ribo")+
        ggtitle("Percent ribosomal")
p6 <- FeaturePlot(so.stromal, features = "soupX_fraction", reduction = "umap")+
    scale_colour_gradientn(colours = viridis(256, option = "D"), name = "soupX_fraction")+
        ggtitle("SoupX fraction")
p <- plot_grid(p1, p2, p3, p4, p5, p6,ncol = 2)

ggsave(paste0(directory, "/Annotation_figures/QC_metrics_UMAP.png"), 
       p, width = 10, height = 12, bg = "white")
p

```

### Markers

Read in the markers calculated by the script "Subset_recluster.Rmd" from 20250108_14-06_Subset_Adipocyte.VEC.LEC.Mural.dir 

```{r, message = FALSE}
markers <- read.table(paste0("20250114_15-14_Subset_Adipocyte.VEC.LEC.Mural.dir/Marker_lists/Markers_", clustering, "_X_scVI_snn_res.", resolution, ".txt"), 
                sep = "\t")

```



### Plot groups of known markers

These groups of markers are defined from the literature

Define the plotting function

```{r}
plot_dotplot <- function (so, genes, category, resolution, fig_width = 7, fig_height = 7){
    p <- DotPlot(so, 
             features = genes,
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle(paste0(category, " markers: resolution ", resolution)) + 
    theme(plot.title = element_text(size = 14, face = "bold"))+
    coord_flip()
    
    ggsave(paste0(directory, "/Marker_dotplots/", category, "_", resolution, ".png"), bg = "white")
    p
}


```

###  general markers

```{r, fig.width=10, fig.height=12}
general_markers <- c("COL1A1", "COL1A2", "COL3A1", "DCN",
              "HMCN1", "NEGR1", "DCLK1", "KAZN", "BICC1", "PRICKLE1",
              "CLU", "COMP", "MMP3", "NOVA1", "CILP", "FBLN1", "PRG4",
              "F13A1", "CD163", "MRC1", "MSR1",
              "PECAM1", "PTPRB", "FLT1", "VWF",
              "NOTCH3", "PDGFRB", "MYO1B",
              "TRDN", "TTN", "NEB", "TNNT1", "LGR5", "ATP2A2",
              "PLIN1", "AQP7", "ADIPOQ",
              "CD247", "SKAP1", "THEMIS",
              "TNNT3", "MYH1", "MYH3",
              "POSTN", "CSMD2", "ADAM12", "FN1", "CEMIP",
              "HLA-DRA", "CD86", "CD74",
              "ASPM", "DIAPH3", "TOP2A",
              "MMRN1", "PROX1", "PKHD1L1",
              "FCN1", "LYZ", "AOAH",
              "BCL11A", "CUX2", "CLEC4C",
              "TENM2", "PTCH2", "APOD", "CNTN4", "ABCA10",
               "KIT", "CPA3", "IL18R1",
              "AK5", "ACP5", "MMP9")
plot_dotplot(so.stromal, general_markers, "general", resolution, fig_width = 10, fig_height = 12)
```



## Remove cluster 7 and 10 as soup

```{r}
so.stromal.subset <- subset(so.stromal, idents = c("7", "10"), invert = TRUE)
DimPlot(so.stromal.subset, reduction = "umap")
```
Is it then worth reclustering?

```{r}
so.stromal.subset <- so.stromal.subset %>% 
           NormalizeData() %>%
           FindVariableFeatures() %>% 
           ScaleData() %>% 
           RunPCA()

# re-calculate neighbours using the scVI embeddings
so.stromal.subset <- FindNeighbors(so.stromal.subset, reduction = "X_scVI", 
                               dims = 1:ncol(Embeddings(so.stromal.subset, "X_scVI")), # use all the dims that were calc by scVI
                               graph.name = c("X_scVI_nn", "X_scVI_snn"))

# recalculate the UMAP using scVI embedidngs
so.stromal.subset <- RunUMAP(so.stromal.subset, reduction = "X_scVI", dims = 1:ncol(Embeddings(so.stromal.subset, "X_scVI")))

```


```{r, fig.width=10, fig.height=10}
so.stromal.subset <- FindClusters(so.stromal.subset, resolution = seq(0.1, 0.6, 0.1), graph.name = "X_scVI_snn")
clustree(so.stromal.subset, prefix = "X_scVI_snn_res.", node_colour = "sc3_stability") + 
        ggtitle("Clustree")
ggsave(paste0(directory, "/Subset_figures/Clustree_plot_subset.png"), 
       width = 10, height = 10, bg = "white")
```

Plot UMAP for each clustering resolution

```{r, fig.width=10, fig.height=16}

resolutionList <- grep("^X_scVI_snn_res", colnames(so.stromal.subset@meta.data), value = TRUE)

plot_list <- list()

for (resolution in resolutionList){
      plot_list [[resolution]] <- DimPlot(object = so.stromal.subset, label = TRUE, reduction = "umap", 
                                          group.by = resolution, shuffle = TRUE)
      }

title <- ggdraw() + draw_label("UMAPs of multiple clustering resolutions", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, ncol = 2) 
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.05, 1))

ggsave(paste0(directory, "/Subset_figures/UMAP_compare_resolutions_subset.png"), 
       width = 10, height = 16, bg = "white")

p

```

Find markers at res 0.4

```{r}
Idents(so.stromal.subset) <- so.stromal.subset$X_scVI_snn_res.0.4
markers_subset_0.4 <-  FindAllMarkers(so.stromal.subset,
                           assay = "soupX",
                           only.pos = TRUE, 
                           min.pct = 0.25, 
                           logfc.threshold = 0.25)
top10_markers_subset_0.4 <- markers_subset_0.4 %>% 
          group_by(cluster) %>% 
          slice_max(n=10, order_by = abs(avg_log2FC))
```
```{r, fig.width=10, fig.height=10}

DoHeatmap(so.stromal.subset, 
                   features = unique(top10_markers_subset_0.4$gene),
                   assay = "soupX") + 
            scale_fill_viridis()
dir.create(paste0(directory, "/Heatmaps"))
ggsave(paste0(directory, "/Heatmaps/Top10markers_subset_0.4.png"), width = 12, height = 12)
```


Happy to go ahead with the new subsetted version so replace the subset back to the normal Seurat Object.

Try with resolution 0.4 first, might end up merging some clusters though.

```{r}
so.stromal <- so.stromal.subset
rm(so.stromal.subset)
resolution <- 0.4
so.stromal$soupX_snn_res.0.4_subset <- Idents(so.stromal)
```


### UMAP split by patient

```{r, fig.height=8, fig.width=10}
DimPlot(object = so.stromal, reduction = "umap", shuffle = TRUE, 
             split.by = "patient", ncol = 3, label = TRUE)+
    theme(panel.background = element_rect(colour = "black", linewidth = 1))+
    ggtitle(paste0("UMAP split by patient at resolution ", resolution))
ggsave(paste0(directory, "/Annotation_figures/UMAP_patient_subset.png"), 
       width = 10, height = 7, bg = "white")
```
### What % of nuclei are from each patient?

```{r}
# make a df of patient percentages per cluster
df <- so.stromal[[]] %>% select(seurat_clusters, patient)
df <- data.table(df)
df <- df[, .(COUNT = .N), by = names(df)]
df2 <- df %>% pivot_wider(names_from = patient, values_from = COUNT, values_fill = 0) %>% 
    column_to_rownames(var = "seurat_clusters")%>%
    mutate_at(vars(1:6), as.numeric) 
df2$my_row_sum <- rowSums(df2)
df2 <- df2 %>% mutate(across(where(is.numeric), ~ .x/my_row_sum*100))
df2$my_row_sum <- NULL
df2$seurat_clusters <- rownames(df2)

df2

# plot as bar chart and heatmap
df3 <- df2 %>% pivot_longer(1:6, names_to = "patient", values_to = "proportion")

ggplot(df3, aes(fill=patient, y=proportion, x=seurat_clusters)) + 
    geom_bar(position="stack", stat = "identity")+
    theme_classic()
ggsave(paste0(directory, "/Annotation_figures/Patient_proportion_barchart_", resolution, ".png"), 
       width = 8, height = 6, bg = "white")

ggplot(df3, aes(y=seurat_clusters, x=patient, fill = proportion))+
    geom_tile()+
    scale_fill_viridis()
ggsave(paste0(directory, "/Annotation_figures/Patient_proportion_heatmap_subset", resolution, ".png"), 
       width = 8, height = 6, bg = "white")


```
No one cluster is >80% from one patient
Cluster 7 is predominantly MSK1556

### Visualise QC metrics 

```{r, message = FALSE, fig.width=10, fig.height=18}
p1 <- VlnPlot(so.stromal, features = "sum", assay = "RNA", pt.size = 0)+
    scale_y_log10() +
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("nCounts")

p2 <- VlnPlot(so.stromal, features = "detected", assay = "RNA", pt.size = 0)+
    scale_y_log10() +
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("nFeatures")

p3 <- VlnPlot(so.stromal, features = "subsets_mito_percent", assay = "RNA", pt.size = 0)+
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("Percent mitochondrial reads")

p4 <- VlnPlot(so.stromal, features = "decontX_contamination", assay = "RNA")+
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("decontX score")

p5 <- VlnPlot(so.stromal, features = "percent.ribo", assay = "RNA")+
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("Percent ribosomal")

p6 <- VlnPlot(so.stromal, features = "soupX_fraction", assay = "RNA")+
    scale_fill_viridis_d()+
    theme(legend.position="none")+
    theme_cowplot()+
    ggtitle("SoupX fraction")

p <- plot_grid(p1, p2, p3, p4, p5, p6, ncol = 1)

ggsave(paste0(directory, "/Annotation_figures/QC_metrics_across_clusters_subset.png"), 
       p, width = 10, height = 21, bg = "white")
p

```

Plot QC metrics on UMAP

```{r, fig.height=12, fig.width=10, message = FALSE}
p1<- FeaturePlot(so.stromal, features = "sum", reduction = "umap")+ggtitle ("nCounts")+
    scale_color_continuous(type = "viridis", limits = c(0,10000))
p2<- FeaturePlot(so.stromal, features = "detected", reduction = "umap") + ggtitle("nFeatures")+
    scale_color_continuous(type = "viridis", limits = c(0,5000))
p3<- FeaturePlot(so.stromal, features = "subsets_mito_percent", reduction = "umap") + ggtitle ("Percent mito reads")+
    scale_color_continuous(type = "viridis")
p4<- FeaturePlot(so.stromal, features = "decontX_contamination", reduction = "umap")+
    scale_colour_gradientn(colours = viridis(256, option = "D"), name = "decontX score")+
        ggtitle("decontX_contamination")
p5 <- FeaturePlot(so.stromal, features = "percent.ribo", reduction = "umap")+
    scale_colour_gradientn(colours = viridis(256, option = "D"), name = "percent.ribo")+
        ggtitle("Percent ribosomal")
p6 <- FeaturePlot(so.stromal, features = "soupX_fraction", reduction = "umap")+
    scale_colour_gradientn(colours = viridis(256, option = "D"), name = "soupX_fraction")+
        ggtitle("SoupX fraction")
p <- plot_grid(p1, p2, p3, p4, p5, p6,ncol = 2)

ggsave(paste0(directory, "/Annotation_figures/QC_metrics_UMAP_subset.png"), 
       p, width = 10, height = 12, bg = "white")
p

```

## Plot general markers

General markers

```{r, fig.width=10, fig.height=12}
plot_dotplot(so.stromal, general_markers, "general_markers", resolution, fig_width = 10, fig_height = 12)
```

## Top markers dotplot with dendogram

scDotPlot
https://bioconductor.org/packages/devel/bioc/vignettes/scDotPlot/inst/doc/scDotPlot.html

Use the sce syntax. Seurat version not working, probably as am using seurat v4 not 5. 
```{r}

# convert to sce
sce <- as.SingleCellExperiment(so.stromal)

# select the top 10 markers for each cluster
features_10 <- markers_subset_0.4 %>% group_by(cluster) %>%
    dplyr::slice(1:10) %>%
    dplyr::select(cluster, gene)

# remove duplicated rows
features_10 <- features_10[!duplicated(features_10$gene),]

# make features into a vector
features_10 <- features_10 %>%
    deframe()

# select the top 5 markers for each cluster
features_5 <- markers_subset_0.4 %>% group_by(cluster) %>%
    dplyr::slice(1:5) %>%
    dplyr::select(cluster, gene)

# remove duplicated rows
features_5 <- features_5[!duplicated(features_5$gene),]

# make features into a vector
features_5 <- features_5 %>%
    deframe()

# select the top 20 markers for each cluster
features_20 <- markers_subset_0.4 %>% group_by(cluster) %>%
    dplyr::slice(1:20) %>%
    dplyr::select(cluster, gene)

# remove duplicated rows
features_20 <- features_20[!duplicated(features_20$gene),]

# make features into a vector
features_20 <- features_20 %>%
    deframe()

# add marker genes to row data
rowData(sce)$Marker_10 <- features_10[match(rownames(sce), features_10)] %>% 
    names()
rowData(sce)$Marker_5 <- features_5[match(rownames(sce), features_5)] %>% 
    names()
rowData(sce)$Marker_20 <- features_20[match(rownames(sce), features_20)] %>% 
    names()

```

Plot 10 most variable genes 

```{r, fig.width=8, fig.height=12}
ident_name <- "X_scVI_snn_res.0.4"
n_colours <- nrow(unique(so.stromal[[ident_name]]))

# plot the log counts
sce %>%
    scDotPlot(features = features_10,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_10",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_10" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_logcounts_10.png"), width = 8, height = 12, bg = "white")

# plot the scaled data
sce %>%
    scDotPlot(features = features_10,
              scale = TRUE,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_10",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_10" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_scaled_10.png"), width = 8, height = 12, bg = "white")

```

2/5 (adipocytes) look nearly identical, think it is the soupX score that has driven the separation of 5
0/1/7 similar VEC
3/4/8 similar  but distinct
6 LEC
9 distinct maybe muscel or fb?


Plot 5 most variable genes 

```{r, fig.width=8, fig.height=8}

# plot the log counts
sce %>%
    scDotPlot(features = features_5,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_5",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_5" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_logcounts_5.png"), width = 8, height = 8, bg = "white")

# plot the scaled data
sce %>%
    scDotPlot(features = features_5,
              scale = TRUE,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_5",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_5" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_scaled_5.png"), width = 8, height = 8, bg = "white")

```

Plot 20 most variable genes 

```{r, fig.width=8, fig.height=16}

# plot the log counts
sce %>%
    scDotPlot(features = features_20,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_20",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_20" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_logcounts_20.png"), width = 8, height = 16, bg = "white")

# plot the scaled data
sce %>%
    scDotPlot(features = features_20,
              scale = TRUE,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_20",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_20" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_scaled_20.png"), width = 8, height = 16, bg = "white")

```




## literature markers





## Adipocytes markers


Markers cited here: https://www.biorxiv.org/content/10.1101/2023.09.04.555678v1.full

Adipocytes key markers
ADIPOQ
AQP7
LPL
PLIN1

Pre-adipocytes
PDGFRA
ASPC1 DPP4, CD55
ASPC2 PPARG
ASPC3 ITGB1, THY1, CD9
ASPC4 PDE4D
ASPC5 ALDH1A3


Specialised adipocytes
No markers given

```{r}
adipocyte_markers <- c("ADIPOQ", "AQP7", "LPL", "PLIN1")
plot_dotplot(so.stromal, adipocyte_markers, resolution = resolution, category = "general_adipoctye_markers")
```
2 definitely adipocytes
5 less clear but probably 

```{r}
adipocyte_precursors <- c("PDGFRA", "DPP4", "CD55", "PPARG", "ITGB1", "THY1", "CD9", "PDE4D", "ALDH1A3")
plot_dotplot(so.stromal, adipocyte_precursors, resolution = resolution, category = "adipocyte_precursors")
```
No obvious clusters of precursors (no PDGRFA expression)


Are top cluster markers from Lazarescu expresse in my data?

Create a list of top 50 DE genes for each cluster from Lazarescu et al and do a dotplot

```{r, fig.width=24, fig.height=20, message=FALSE, warning=FALSE}
markers.laz <- read.table("Files.dir/Markers_Lazarescu.txt", header = TRUE, sep = "\t")

plot_list <- list()
laz_top50 <- list()
for (i in unique(markers.laz$cluster)){
    print(i)
    laz_top50[[i]] <- markers.laz %>% 
        group_by(cluster) %>% 
        slice_max(n=50, order_by = abs(avg_log2FC)) %>% 
        filter(cluster == i) %>%
        pull (gene)
    plot_list[[i]] <- plot_dotplot(so.stromal, laz_top50[[i]], paste0(i, "_50"), resolution)
    
}
    
# Plot all dotplots together
plot_grid(plotlist = plot_list, ncol = 3)
ggsave(paste0(directory, "/Marker_dotplots/Lazarescu_top50.png"), width = 20, height = 20, bg = "white")
```
Not informative


Plot other groups of markers

## My Immune markers

```{r, fig.width=12, fig.height=12}
plot_immune_dotplots <- function (so, resolution){
    
    p1 <- DotPlot(so, 
             features =c("PTPRC", "CD247","IL7R", "CCR7", "THEMIS"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("T cell markers") + 
    coord_flip()
    
    p2 <- DotPlot(so, 
             features =c("MS4A1", "CD37", "BLNK", "CD38", "CD79A", "CD79B", "SDC1", "CD27", "CD78"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("B cell markers") + 
    coord_flip()
    
    p3 <- DotPlot(so, 
             features =c("CD14", "CD163", "CD163L1", "MERTK", "MRC1", "MSR1"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Macrophage markers") + 
    coord_flip()
    
    p4 <- DotPlot(so, 
             features =c("CD300A", "IRF7", "CLEC4C", "IL3RA", "NRP1", "HLA-DQA1", "CD74", "GPAT3", "CIITA", "HLA-DRB1"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Dendritic cell markers") + 
    coord_flip()
    
    p5 <- DotPlot(so, 
             features =c("KIT", "CDK15", "MS4A2", "IL18R1", "HPGD"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Granulocyte_markers") + 
    coord_flip()
    
    p6 <- DotPlot(so, 
             features =c("CD14", "VCAN", "NCF1", "MS4A6A", "FCN1",  # classical monocyte
                         "FCGR3A", "LST1", "RPS19", "MS4A7", "CSF1R", "CDKN1C", "LILRB2", "ITGAL"), #non-classical monocytes
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Monocyte_markers") + 
    coord_flip()
    
    #add an overall title and arrange the graphs on one page
    title <- ggdraw() + draw_label(paste0("Immune cell markers at resolution ", resolution), fontface='bold', size = 16)
    p <- plot_grid(p1, p2, p3, p4, p5, p6, ncol = 2)
    p <- plot_grid(title, p, ncol=1, rel_heights=c(0.05, 1))
    
    p

    
}

plot_immune_dotplots(so.stromal, resolution)
ggsave(paste0(directory, "/Marker_dotplots/dotplot_immune_cell_markers_", resolution, ".png"), 
       width = 12, height = 12, 
       bg = "white")
```

No immune cells at this resolution


```{r, fig.height=20, fig.width=12}
plot_muscle_markers <- function (so, resolution){
    p1 <- DotPlot(so, 
             features =c("TRDN", "DES", "ACTN2"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("General Muscle markers") + 
    coord_flip()
    
    p2 <- DotPlot(so, 
             features =c("TNNC2", "TNNI2", "TNNT3", "ACTN3", "HOMER1", "KCNQ5", "MYH1", "MYH3", "MYLK4", "MYBPC2"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Fast twitch markers") + 
    coord_flip()
    
    p3 <- DotPlot(so, 
             features =c("COL22A1", "ADAMTSL1", "CPM", "CSMD1", "FRAS1", "PZP", "RALYL", "SAMD5", "SORBS2", "SPAG16"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Transitional markers") + 
    coord_flip()
    
    p4 <- DotPlot(so, 
             features =c("TNNC1", "TNNI1", "TNNT1", "ATP2A2", "CCR3", "LGR5", "MYH7", "MYH7B", "NEK10", "TECRL"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Slow twitch markers") + 
    coord_flip()
    
    p5 <- DotPlot(so, 
             features =c("ACTA2", "MYH11", "TAGLN", "CRIP2", "MYL9", "RTL8C", "RERGL", "PLAC9", "MFAP4", "LGALS3BP"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Smooth muscle markers") + 
    coord_flip()
    
    p6 <- DotPlot(so, 
             features =c("PAX7", "CALCR", "CDH4", "CLCN5", "CTNND2", "GREM1"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Satellite cells markers") + 
    coord_flip()
        
 
    #add an overall title and arrange the graphs on one page
    title <- ggdraw() + draw_label(paste0("Muscle cell markers at resolution ", resolution), fontface='bold', size = 16)
    p <- plot_grid(p1, p2, p3, p4, p5, p6, nrow = 3)
    p <- plot_grid(title, p, ncol=1, rel_heights=c(0.05, 1))
    
    p
}

plot_muscle_markers(so.stromal, resolution)
ggsave(paste0(directory, "/Marker_dotplots/Muscle_markers.png"), width = 12, height = 20, bg = "white")

```

7 contains some muscle markers, again this could be the high soupX content. 
4 is likely smooth muscle. 


```{r, fig.width=12, fig.height=16}
plot_dotplots <- function (so, resolution){
    
    p1 <- DotPlot(so, 
             features =c("COL1A1", "COL1A2", "COL3A1", "COL3A2",  "DCN", "PRG4"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Fibroblast markers") + 
    coord_flip()
    
    p2 <- DotPlot(so,
             features =c("PECAM1", "PTPRB", "VWF", "ADGRL4", "CD34", "CDH5", "FLT1", "NOTCH4", "TEK"),
             assay = "soupX")  + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Vascular endothelial markers") + coord_flip()
    
    p3 <- DotPlot(so,
             features =c("EPHB4", "FLT4", "KDR", "LYVE1", "MMRN1", "NRG3", "PKHD1L1", "PROX1"),
             assay = "soupX")  + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Lymphatic endothelium markers") + coord_flip()
    
    p4 <- DotPlot(so, 
             features =c("TRDN", "DES", "ACTN2", "TNNC2", "TNNI2", "TNNT3", "TNNC1", "TNNI1", "TNNT1"),
             assay = "soupX")  + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Skeletal muscle markers") + 
    coord_flip()
    
    p5 <- DotPlot(so, 
             features =c("ABCA8", "COL15A1", "CD55", "DCLK1", "ELN", "FBN1", "FBLN1", 
                         "FBLN2", "FBLN5", "KCND2", "NEGR1", "NOVA1", "PDGFRA", "VIT"),
             assay = "soupX")  + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Mural cell markers") + 
    coord_flip()
    
   p6 <- DotPlot(so, 
             features =c("GPAM", "TMEM132C", "PLIN1", "PLIN4", "AQP7", "ADIPOQ", "PDE3B", "FABP4", "LIPE"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Adipocyte markers") + 
    coord_flip()
    
    p7 <- DotPlot(so, 
             features =c("NTRK3", "TENM2", "KANK4", "SLC2A1", "SLC22A3", "CLDN1"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Nerve-like cell markers") + 
    coord_flip()
    
    p8 <- DotPlot(so, 
             features =c("PAX7", "CALCR", "CDH4", "CLCN5", "CTNND2", "GREM1"),
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle("Satellite cell markers") + 
    coord_flip()

    #add an overall title and arrange the graphs on one page
    title <- ggdraw() + draw_label(paste0("Cell markers at resolution ", resolution), fontface='bold', size = 16)
    p <- plot_grid(p1, p2, p3, p4, p5, p6, p7, p8, ncol = 2)
    p <- plot_grid(title, p, ncol=1, rel_heights=c(0.05, 1))
    
    p
    
    
}

plot_dotplots(so.stromal, resolution)
```

Finer annotation of endothelium from Quads paper


```{r, fig.width=12, fig.height=10}
endothelial_markers <- list("Pericytes" = c("NOTCH3", "PDGFRB", "MYO1B", "DLC1", "CALD1", "EBF2", "COL6A3",
                                            "COL25A1", "POSTN", "STEAP4", "RGS5"),
                            "VECs" = c("PCAM1", "VWF", "PTPRB", "FLT1", "EMCHN"),
                            "Venular VECs" = c("SLC02A1", "SELP", "NOSTRIN", "MYRIP", "GNA14", "TACR1", "RAMP3", "LIFR", "ICAM1"),
                            "Arteliolar VECs" = c("PODXL", "EFNB2", "NEBL", "SYT1", "THSD4", "CXCL12", "SEMA3G", "GJA4", "HEY1"),
                            "Capillary VECs" = c("COL4A1", "COL4A2", "COL15A1", "DYSF", "ARHGAP18", "LAMB1", "SPARC", "GNG2", "VWA1"),
                            "LECs" = c("PROX1", "MMRN1", "TFP1", "RELN", "PKJD1L1", "SEMA3A", "FLRT2", "CCL21", "LYVE1", "STAB2"),
                            "vSMCs" = c("MYH11", "TAGLN", "AGRL3","SORBS2", "RCAN2", "MYOD", "LMOD1", "CARMN")
                            #"Dividing VEC" = c("ASPM", "DIAPH3", "TOP2A", "CENPE", "TPX2", "MELK")
                           )

DotPlot(so.stromal, endothelial_markers, assay = "soupX")+
    scale_colour_viridis(direction = -1) +
    ggtitle(paste0("Endothelial markers: resolution ", resolution)) + 
    theme(plot.title = element_text(size = 14, face = "bold"))+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

Cluster 7 doesn't fit well and in fact is mostly from sample MSK1556. 
It is most similar to cluster 0 so should be merged with 0. 



More detailed markers for mural cells from: https://www.cell.com/developmental-cell/fulltext/S1534-5807(22)00685-2
Data from: http://betsholtzlab.org/Publications/SMC/database.html
Cluster 3/8 Arterial (heart & colon) SMC


```{r}
arterial_SMC_markers <- c("OLFR558", "CHCHD10", "OLFR78", "TESC", "SERPINI1", "EPHX3", 
                          "PLN", "SORBS2", "SNCG", "SLC38A11", "NRIP2", "PERP", "NDUFA412", 
                          "RGS4", "COX412", "RGS5", "GJA4", "CASQ2", "NGF", "PLCE1", "LPL", 
                          "TMEM41", "GPR20", "GPCPD1", "UBA2", "FGF1")

venous_SMC_markers <- c("COLEC11", "IGFBP3", "CCL11", "SDC1", "CHRDL1", "COL14A1", "CYGB", 
                        "IFIM1", "ALDH1A1", "DP", "VCAM1", "AR4", "GPX3", "LY6E", "REM1", 
                        "SLC43A3", "THBS2", "PRRX1", "MATN2", "LASP1", "IGFBP4", "SEPT4", 
                        "DCN", "ANGPL4", "OAF", "GPC3", "SRPX") 

plot_dotplot(so.stromal, arterial_SMC_markers, "Arterial_SMC", resolution)
plot_dotplot(so.stromal, venous_SMC_markers, "Venous_SMC", resolution)
```
As expected, not super helpful as they come from mouse single cell. 

https://www.nature.com/articles/nature25739
Pericytes PDGFRB, CSPG4, ANPEP, RGS5, CD248, ABCC9, VTN, S1PR3
arterial SMC CNN1 ACTA2 tagln
Broad mural CSPG4, PDGFRb

There are markers in supp data but there are only pariwise comparisons between cell types, not helpful!


```{r}
percityes_van <- c("PDGFRB", "CSPG4", "ANPEP", "RGS5", "CD248", "ABCC9", "VTN", "S1PR3")
mural_arterial_SMC_van <- c("CSPG4", "PDGFRB", "CNN1", "ACTA2", "TAGLN")

plot_dotplot(so.stromal, percityes_van, "Pericytes_van", resolution)
plot_dotplot(so.stromal, mural_arterial_SMC_van, "Mural_Arterial_SMC_van", resolution)
```

Cluster 8 pericytes (more than 3 or 4)

Plot at slightly lower resolution where 3 and 4 are merged

```{r, fig.width=12, fig.height=10}
Idents(so.stromal) <- so.stromal$X_scVI_snn_res.0.1
DotPlot(so.stromal, endothelial_markers, assay = "soupX")+
    scale_colour_viridis(direction = -1) +
    ggtitle(paste0("Endothelial markers: resolution 0.1")) + 
    theme(plot.title = element_text(size = 14, face = "bold"))+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```



# Summary of markers

Use this code to get a list of markers for each cluster

```{r}

cluster_markers <- list()
for (i in levels(Idents(so.stromal))){
    cluster_markers[[i]] <- top10_markers_subset_0.4 %>% 
        group_by(cluster) %>% 
        slice_max(n=20, order_by = abs(avg_log2FC)) %>% 
        filter(cluster == i) %>% pull (gene)
}
cluster_markers

write.csv(t(as.data.frame(cluster_markers)), paste0(directory, "/Marker_list_for_cellxgene_subset.csv"), quote = FALSE)

```


9 looks like skeletal muscle on cellxgene






0 - venular VEC
1 - Arteriolar VEC
2 - adipocyte
3 - vSMC
4 - vSMC
5 - adipocytes high soupX
6 - LEC
7 - VEC from MSK1556
8 - pericytes?
9 - skeletal muscle cells


2/5 merge as too similar adipocytes
0/7 merge as patient specific
3/4 merge (they may in fact be different types of SMC but not sure the literature is strong enough to support this)

## Rename identities

```{r, fig.width=10, fig.height=7}

so.stromal <- RenameIdents(so.stromal, 
                           `0` = "Venular VEC",
                          `1` = "Arteriolar VEC",
                          `2` = "Adipocytes",
                          `3` = "vSMC", 
                          `4` = "vSMC",
                          `5` = "Adipocytes",
                          `6` = "Lymphatic endothelial cells",
                          `7` = "Venular VEC",
                          `8` = "Pericytes",
                          `9` = "Skeletal muscle cells")
so.stromal$fine_annotation_stromal <- Idents(so.stromal)

DimPlot(so.stromal, reduction = "umap", label = TRUE, repel = TRUE)
ggsave(paste0(directory, "/Annotation_figures/UMAP_annotated.png"), width = 6, height = 4, bg = "white")

```

Project this on the original umap

```{r, fig.width=10, fig.height=7}
DimPlot(so.stromal, reduction = "umap.scvi", label = TRUE, repel = TRUE)
ggsave(paste0(directory, "/Annotation_figures/UMAP_annotated_original_dims.png"), width = 6, height = 4, bg = "white")
```

## Find Markers

```{r}

markers <-  FindAllMarkers(so.stromal,
                           assay = "soupX",
                           only.pos = TRUE, 
                           min.pct = 0.25, 
                           logfc.threshold = 0.25
                        )
write.table(markers, 
            paste0(directory, "/Markers_annotated.txt"), 
                quote = FALSE, sep = "\t")
```

# Summary of markers

Use this code to get a list of markers for each cluster

```{r}

cluster_markers <- list()
for (i in levels(Idents(so.stromal))){
    cluster_markers[[i]] <- markers %>% 
        group_by(cluster) %>% 
        slice_max(n=20, order_by = abs(avg_log2FC)) %>% 
        filter(cluster == i) %>% pull (gene)
}
cluster_markers

write.csv(t(as.data.frame(cluster_markers)), paste0(directory, "/Marker_list_for_cellxgene_annotated.csv"), quote = FALSE)

```

## How to name the clusters?

Do FeaturePlots and DotPlots of the top 9 markers for each cluster

```{r, fig.width=10, fig.height=10}
dir.create(paste0(directory, "/Feature_Plots_annotated"))
dir.create(paste0(directory, "/Top_20_Dotplots_annotated"))
for (i in levels(Idents(so.stromal))){
    FeaturePlot(so.stromal, cluster_markers[[i]][1:9], reduction = "umap")
    ggsave(paste0(directory, "/Feature_Plots_annotated/Top9_cluster_", i, ".png"), width = 10, height = 10, bg = "white")
    
    DotPlot(so.stromal, 
             features = cluster_markers[[i]],
             assay = "soupX") + 
    scale_colour_viridis(direction = -1) +
    ggtitle(paste0("Top 20 markers cluster", i)) + 
    theme(plot.title = element_text(size = 14, face = "bold"))+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+coord_flip()
    ggsave(paste0(directory, "/Top_20_Dotplots_annotated/Top20_cluster_", i, ".png"), width = 7, height =7, bg = "white")
}
```


```{r, fig.width=10, fig.height=10}

top10_markers <-  markers %>% 
          group_by(cluster) %>% 
          slice_max(n=10, order_by = abs(avg_log2FC))


DoHeatmap(so.stromal, 
                   features = unique(top10_markers$gene),
                   assay = "soupX") + 
            scale_fill_viridis()
dir.create(paste0(directory, "/Heatmaps"))
ggsave(paste0(directory, "/Heatmaps/Top10markers_annotated.png"), width = 12, height = 12)
```


UMAP split by microanatomy

```{r, fig.height=8, fig.width=8}
p <- DimPlot(object = so.stromal, reduction = "umap", shuffle = TRUE, 
             split.by = "microanatomical_site", ncol = 2)+
    ggtitle(paste0("UMAP split my microanatomy"))
ggsave(paste0(directory, "/Annotation_figures/UMAP_by_microanatomy_annotated_", resolution, ".png"), 
       width = 8, height = 8, bg = "white")
p
```
Bar chart of cell composition

```{r, fig.width = 8, fig.height=6}

df <- so.stromal[[]] %>% select(fine_annotation_stromal, microanatomical_site)

#library(data.table)
df <- data.table(df)
df <- df[, .(COUNT = .N), by = names(df)]

ggplot(df, aes(fill=microanatomical_site, y=COUNT, x=fine_annotation_stromal)) + 
    geom_bar(position="fill", stat="identity")+
    theme_classic()+
    scale_fill_manual(values = ma.cols)+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

ggsave(paste0(directory, "/Annotation_figures/Cluster_fraction_microanatomy_annotated.png"), 
       width = 8, height = 6, bg = "white")
```

### What % of nuclei are from each patient?

```{r}
# make a df of patient percentages per cluster
df <- so.stromal[[]] %>% select(seurat_clusters, patient)
df <- data.table(df)
df <- df[, .(COUNT = .N), by = names(df)]
df2 <- df %>% pivot_wider(names_from = patient, values_from = COUNT, values_fill = 0) %>% 
    column_to_rownames(var = "seurat_clusters")%>%
    mutate_at(vars(1:6), as.numeric) 
df2$my_row_sum <- rowSums(df2)
df2 <- df2 %>% mutate(across(where(is.numeric), ~ .x/my_row_sum*100))
df2$my_row_sum <- NULL
df2$seurat_clusters <- rownames(df2)

df2

# plot as bar chart and heatmap
df3 <- df2 %>% pivot_longer(1:6, names_to = "patient", values_to = "proportion")

ggplot(df3, aes(fill=patient, y=proportion, x=seurat_clusters)) + 
    geom_bar(position="stack", stat = "identity")+
    theme_classic()
ggsave(paste0(directory, "/Annotation_figures/Patient_proportion_barchart_", resolution, ".png"), 
       width = 8, height = 6, bg = "white")

ggplot(df3, aes(y=seurat_clusters, x=patient, fill = proportion))+
    geom_tile()+
    scale_fill_viridis()
ggsave(paste0(directory, "/Annotation_figures/Patient_proportion_heatmap_", resolution, ".png"), 
       width = 8, height = 6, bg = "white")


```

### How many cells are there per cluster?

```{r}
cell_numbers <- table(Idents(so.stromal)) %>% as.data.frame()
colnames(cell_numbers) <- c("cluster", "frequency")
cell_numbers
```

## Top markers dotplot with dendogram

scDotPlot
https://bioconductor.org/packages/devel/bioc/vignettes/scDotPlot/inst/doc/scDotPlot.html

Use the sce syntax. Seurat version not working, probably as am using seurat v4 not 5. 
```{r}

# convert to sce
sce <- as.SingleCellExperiment(so.stromal)

# select the top 10 markers for each cluster
features_10 <- markers %>% group_by(cluster) %>%
    dplyr::slice(1:10) %>%
    dplyr::select(cluster, gene)

# remove duplicated rows
features_10 <- features_10[!duplicated(features_10$gene),]

# make features into a vector
features_10 <- features_10 %>%
    deframe()

# select the top 5 markers for each cluster
features_5 <- markers %>% group_by(cluster) %>%
    dplyr::slice(1:5) %>%
    dplyr::select(cluster, gene)

# remove duplicated rows
features_5 <- features_5[!duplicated(features_5$gene),]

# make features into a vector
features_5 <- features_5 %>%
    deframe()


# add marker genes to row data
rowData(sce)$Marker_10 <- features_10[match(rownames(sce), features_10)] %>% 
    names()
rowData(sce)$Marker_5 <- features_5[match(rownames(sce), features_5)] %>% 
    names()

```

Plot 10 most variable genes 

```{r, fig.width=8, fig.height=12}
ident_name <- "fine_annotation_stromal"
n_colours <- nrow(unique(so.stromal[[ident_name]]))

# plot the log counts
sce %>%
    scDotPlot(features = features_10,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_10",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_10" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_logcounts_10_annotated.png"), width = 8, height = 12, bg = "white")

# plot the scaled data
sce %>%
    scDotPlot(features = features_10,
              scale = TRUE,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_10",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_10" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_scaled_10_annotated.png"), width = 8, height = 12, bg = "white")

```

Plot 5 most variable genes 

```{r, fig.width=8, fig.height=8}

# plot the log counts
sce %>%
    scDotPlot(features = features_5,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_5",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_5" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_logcounts_5_annotated.png"), width = 8, height = 8, bg = "white")

# plot the scaled data
sce %>%
    scDotPlot(features = features_5,
              scale = TRUE,
              group = ident_name,
              groupAnno = ident_name,
              featureAnno = "Marker_5",
              groupLegends = FALSE,
              annoColors = list(ident_name = pal_d3()(n_colours),
                                "Marker_5" = pal_d3()(n_colours)),
              annoWidth = 0.02)
ggsave(paste0(directory, "/Annotation_figures/Dotplot_with_dendogram_scaled_5_annotated.png"), width = 8, height = 8, bg = "white")

```
# Save the object

```{r}
dir.create(paste0(directory, "/RDS_objects.dir"))
saveRDS(so.stromal, paste0(directory, "/RDS_objects.dir/Achilles_stromal_subset.rds"))
```

```{r}
sessionInfo()
```

