---
title: "MiloR_Cell_proportions-Achilles"
author: "Carla Cohen"
date: "`r Sys.Date()`"
output: html_document
knit: rmarkdown::render
---

# Cell type proportion analysis

Aim: To analyse the how cell proportions vary across microanatomy in the fibroblast subsets

Using MioR as recommended by Bo Sun & Lucy Garner

https://www.nature.com/articles/s41587-021-01033-z  
https://marionilab.github.io/miloR/articles/milo_demo.html  
https://www.bioconductor.org/packages/devel/bioc/vignettes/miloR/inst/doc/milo_gastrulation.html  
https://github.com/MarioniLab/miloR  

## Steps

1. Set up 
- Set up directories, import yml file, set colours  
- Read in Seurat object  

2. Set up the milo object & perform initial calculations
- convert so to sce  
- create a milo object  
- construct KNN graph  
- define representative neighbourhoods & plot  
- compute neighbourhood connectivity  
- count number of cells per neighbourhood  

3.  Set up the experimental design for differential abundance testing  

4.  Perform the differential abundance test  

5. Plot the outputs 
- Inspect the distribution of p values  
- Volcano plot of test results
- Neighbourhood graph on Harmony UMAP projection  
- Beeswarm plot  



### 1. Set up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(Seurat)
library(tidyverse)
library(miloR)
library(SingleCellExperiment)
library(scater)
library(scran)
library(yaml)
library(cowplot)
library(svglite)

set.seed(1)

# make a new output folder for each run, with the date & time in the directory name
date <- Sys.Date() %>% str_replace_all("-", "")
time <- format(Sys.time(), "%X") %>% str_replace_all(":", "-") %>%
    str_sub(1, 5)
directory <- paste0(date,"_", time, "_MiloR_Cell_Proportions-Muscle.dir")
dir.create(directory, showWarnings = FALSE)
dir.create(paste0(directory, "/Figures/"))
dir.create(paste0(directory, "/RDS_objects.dir/"))
dir.create(paste0(directory, "/Results.dir/"))

# read yaml file
ini <- read_yaml("miloR.yml")

# microanatomy colours
Okabe_Ito <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
ma.cols <-  c(Enth = Okabe_Ito[3], MB = Okabe_Ito[5], MTJ = Okabe_Ito[6], muscle = Okabe_Ito[7])

```

Print the parameters used

```{r}
ini
```



Read in Seurat object

```{r}
so.muscle <- readRDS(paste0(ini$so_location, "/RDS_objects.dir/Achilles_muscle_subset.rds"))
so.muscle

```

```{r}

# add Idents column to metadata
so.muscle$cell_annotation_muscle_0.2 <- Idents(so.muscle)

Idents(so.muscle) %>% unique()
```



Subset the seurat object to remove the Enth & MTJ samples

```{r}
Idents(so.muscle) <- so.muscle$microanatomical_site
so.muscle <- subset(so.muscle, idents = c("MTJ", "muscle"))

so.muscle

```

Plot the UMAP split by microanatomy

```{r, fig.width=10, fig.height=5}
Idents(so.muscle) <- so.muscle$cell_annotation_muscle_0.2
DimPlot(so.muscle, split.by = "microanatomical_site", ncol = 2, reduction = "umap")
ggsave(paste0(directory, "/Figures/UMAP_microanatomy.png"), width = 10, height = 4)
```


NB the following text from the script used to generate paper figures: 
https://github.com/MarioniLab/milo_analysis_2020/blob/main/notebooks/mouse_thymus-figures_revision.Rmd  

"Now we are ready to begin the `Milo` workflow, which begins with building a kNN-graph. Prior to this point it is extremely 
important to remove any batch effects, and use a method or tool that returns either the corrected gene expression values, or a 
corrected reduced dimensional space. We require either the corrected gene expression or corrected reduced dimensions to compute 
the distances between cells for graph building and refining neighbourhoods."

So I think here I should be using either batch corrected PCA, or the harmony embeddings.


## 2. Set up the milo object & perform initial calculations  
Done following the vignette as I have done in the script miloR_vignette (on computer not cluster)

```{r}

# Convert to single cell experiment object  
sce <- as.SingleCellExperiment(so.muscle)


# Create a milo object
achilles_milo <- Milo(sce)

# construct KNN graph
achilles_milo <- buildGraph(achilles_milo, k = 10, d = 30, 
                            reduced.dim = ini$dim_reduction)

# define representative neighbourhoods
achilles_milo <- makeNhoods(achilles_milo, 
                                    prop = 0.1, #if ncells<30k use prop=0.1 
                                    # in the paper they use 0.3, and have 90k cells
                            k = 10, d=30, refined = TRUE, reduced_dims = ini$dim_reduction) # paper uses "PCA"
achilles_milo
```

Plot the distribution of neighbourhood sizes, we are looking for avg neighbourhood size > 5 X N_samples
Achilles data has 18 samples
5 x 18 = 90
Maybe need to adjust k and d to increase the neighbourhood sizes. 

```{r}
plotNhoodSizeHist(achilles_milo)

ggsave(paste0(directory, "/Figures/NeighbourhoodSizeHistogram.png"))
ggsave(paste0(directory, "/Figures/NeighbourhoodSizeHistogram.svg"))
```

Compute neighbourhood connectivity
(NB this step can take a few minutes i.e. half an hour)
```{r}
achilles_milo <- calcNhoodDistance(achilles_milo, d=30, 
                                   reduced.dim = ini$dim_reduction)
```

Count how many cells are in each neighbourhood per sample
```{r}
achilles_milo <- countCells(achilles_milo, 
                            meta.data = as.data.frame(colData(achilles_milo)),
                            sample="sample")
head(nhoodCounts(achilles_milo))
```


## 3. Set up experimental design for differential abundance testing


```{r}
achilles_design <- data.frame(colData(achilles_milo))[,c("sample", "patient", "microanatomical_site")]

## Convert batch info from integer to factor
achilles_design$patient <- as.factor(achilles_design$patient) 
achilles_design <- distinct(achilles_design) # keep unique rows
rownames(achilles_design) <- achilles_design$sample

# add a column to represent microanatomical site as an integer 
# 1 = enth, 2 = MB, 3 = MTJ, 4 = muscle
# try adding on a per-centimeter basis  c(2, 4, 9, 14)), makes no difference
#m.df <- data.frame(microanatomical_site = c("MTJ", "muscle"),
#                   microanatomy_integer = c(1, 2))

# add a column of patient + microanatomical site called "condition"
# achilles_design$condition <- paste(achilles_design$patient, achilles_design$microanatomical_site, sep = "_")

#achilles_design <- left_join(achilles_design, m.df)

# make sure the design table is in the same order as the nhoodCounts table
achilles_design <- achilles_design[colnames(nhoodCounts(achilles_milo)), ,drop=FALSE]
table(achilles_design$microanatomical_site)# shows how many samples at each microanatomy

achilles_design

```


## 4. Perform the differential abundance test  
We want to test for differences across microanatomy while accounting for variability between patients. 

rownames(traj_design) <- traj_design$Sample
da_results <- testNhoods(traj_milo, design = ~ Condition, design.df = traj_design)
```{r}
da_results <- testNhoods(achilles_milo, 
                         design = ~ patient + microanatomical_site,
                         design.df=achilles_design, 
                         fdr.weighting="k-distance", # default
                         reduced.dim = ini$dim_reduction
                         )

da_results$Diff <- sign(da_results$logFC) # add a column to show up or down regulation
da_results$Diff[da_results$SpatialFDR > 0.1] <- 0 # made that column 0 if the result is not significant
table(da_results$Diff)
```

This table shows how many nodes show up- and down-regulation  


## 5. Plot the outputs

#### Inspect the distribution of p values

```{r}
ggplot(da_results, aes(PValue)) + geom_histogram(bins=50)
ggsave(paste0(directory, "/Figures/P-values_distribution.png"))
ggsave(paste0(directory, "/Figures/P-values_distribution.svg"))
```

#### Volcano plot of test results
[presumably could use EnhancedVolcano here if preferred]


```{r}
max.lfc <- max(abs(da_results$logFC))

if (ini$subset_muscle == TRUE){
    left_label <- "Higher in enthesis"
    right_label <- "Higher in MTJ"
}else{
    left_label <- "Higher in enthesis"
    right_label <- "Higher in Muscle"
}

ggplot(da_results, aes(x=logFC, y=-log10(SpatialFDR), colour=as.character(Diff))) +
    geom_hline(yintercept=-log10(0.1), lty=2, colour='grey50') +
    geom_point(size=3) +
    annotate("text", x=-2, y=5, label=left_label, size=4) +
    annotate("text", x=2, y=5, label=right_label, size=4) +
    scale_x_continuous(limits=c(-max.lfc-0.1, max.lfc+0.1)) +
    theme_cowplot() +
    scale_colour_manual(values=c("firebrick2", "grey80", "firebrick2")) +
    guides(colour="none") +
    labs(x="log Fold Change", y=expression(paste("-log"[10], " Spatial FDR")))

ggsave(paste0(directory, "/Figures/Volcano_plot.png"), width = 6, height = 6, bg = "white")
ggsave(paste0(directory, "/Figures/Volcano_plot.svg"), width = 6, height = 6, bg = "white")
```
> Conclusion: there are no significant differences between MTJ and muscle
> We cannot compare across all four microanatomies because only 2 patients have cells in the enthesis. 



#### Save the results  

```{r}

# save the results
write.table(da_results, 
            file = paste0(directory, "/Results.dir/Achilles-milo-DA-results.txt"),
            sep = "\t", quote = FALSE, row.names = FALSE)
```


Print session info

```{r}
sessionInfo()
```

